File: sem_2/AAD/using_c/bubble_sort.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

/* generate the random numbers*/
void generater_random(int a[], int n){
    for(int i = 0; i<n; i++)
        a[i]= rand() %1000;
}

/*bubble sort*/
void bubble_sort(int a[], int n)
{
    int temp;
    for(int i= 0; i<n-1;i++){
        for (int j = 0;j<n-1;j++){
            if(a[j]>a[j+1]){
                temp = a[j];
                a[j] = a[j+1];
                a[j+1] = temp;
            }
        }
    }
}
int main()
{
    int a[10000];
    clock_t start,end;
    double time_taken;
    double theoretical_time;
    FILE *fp;

    srand(time(NULL));
    fp = fopen("bubble_time.txt","w");  
    for(int n = 100; n <=10000; n+=100)
    {
        generater_random(a,n);
        start=clock();
        bubble_sort(a,n);
        end=clock();
        time_taken=(double)(end - start)/CLOCKS_PER_SEC;
        theoretical_time=(double)(n*n)*1e-8;
        fprintf(fp,"%d %lf %lf\n",n,time_taken,theoretical_time);
    }

    fclose(fp);
    FILE *gnuplotPipe = popen("gnuplot -persistent", "w");

    fprintf(gnuplotPipe,"set title 'bubble sort time efficiency'\n");
    fprintf(gnuplotPipe,"set xlabel 'Input Size'\n");
    fprintf(gnuplotPipe,"set ylabel 'Time (seconds)'\n");

    fprintf(gnuplotPipe, "set grid\n");
    fprintf(gnuplotPipe, "set term png\n");
    
    /* Sets the output format to PNG image */
    
    fprintf(gnuplotPipe, "set output 'bubble_sort_graph.png'\n");
    
    fprintf(gnuplotPipe,
    "plot 'bubble_time.txt' using 1:2 with linespoints title 'Actual Time', "
    "'bubble_time.txt' using 1:3 with lines lw 2 title 'Theoretical O(n^2)'\n");
    
    /* Plots data from the file bubble_time.txt; using 1:2 means column 1 -> X axis,
       column 2 -> Y axis */
    
    /* Plots the theoretical curve from column 3, lines draws a smooth line,
       lw 2 sets the line width */

    fflush(gnuplotPipe);
    pclose(gnuplotPipe);
    return 0;
}

File: sem_2/AAD/using_python/bubble_sort.py
import random
import time
import matplotlib.pyplot as plt

# Generate random numbers
def generate_random(n):
    return [random.randint(1,1000) for _ in range(n)]

# Bubble Sort
def bubble_sort(arr):
    n = len(arr)
    for i in range(n-1):
        for j in range(n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]

sizes = []
actual_times = []
theoretical_times = []

with open("bubble_time.txt", "w") as fp:
    for n in range(100, 10001, 500):
        arr = generate_random(n)

        start = time.perf_counter()  # records a high-precision timestamp
        bubble_sort(arr)
        end = time.perf_counter()    # records a high-precision timestamp

        time_taken = end - start
        theoretical_time = (n * n) * 1e-8   # 1e-8 is 0.00000001

        fp.write(f"{n} {time_taken} {theoretical_time}\n")

        sizes.append(n)
        actual_times.append(time_taken)
        theoretical_times.append(theoretical_time)


# Plot Graph
plt.figure(figsize=(10,6))

plt.plot(sizes, actual_times, marker='o', label="Actual Time")  # A circle is drawn at each point
plt.plot(sizes, theoretical_times, linewidth=2, label="Theoretical O(n²)")

plt.title("Bubble Sort Time Efficiency")
plt.xlabel("Input Size")
plt.ylabel("Time (seconds)")
plt.grid(True)
plt.legend()

plt.show()